src/net/ossl.cc

```diff
@@ -557,312 +607,560 @@ namespace tls {
-        auto n = session_fn(_ssl.get());
-        auto ssl_err = SSL_get_error(_ssl.get(), n);
-        switch (ssl_err) {
-        case SSL_ERROR_NONE:
-            break;
-        case SSL_ERROR_WANT_READ:
-            return want_read_fn();
-        case SSL_ERROR_WANT_WRITE:
-            return wait_for_input();
-        case SSL_ERROR_SSL:
-        {
-            // Catch-all for handshake errors
-            auto ec = ERR_GET_REASON(ERR_get_error());
-            switch (ec) {
-            case SSL_R_PEER_DID_NOT_RETURN_A_CERTIFICATE:
-            case SSL_R_CERTIFICATE_VERIFY_FAILED:
-            case SSL_R_NO_CERTIFICATES_RETURNED:
-                verify(); // should throw
-                [[fallthrough]];
-            default:
-                _error = std::make_exception_ptr(ossl_error("Failed to establish SSL handshake"));
-                return make_exception_future<>(_error);
-            }
+    // Called after locking the _in_sem and _out_sem semaphores.
+    // This function will walk through the handshake with a remote peer
+    // If EOF is encountered, ENOTCONN is thrown
+    future<> do_handshake() {
+        if (eof()) {
+            // if we have experienced and eof, set the error and return
+            // GnuTLS will probably return GNUTLS_E_PREMATURE_TERMINATION
+            // from gnutls_handshake in this situation.
+            _error = std::make_exception_ptr(std::system_error(
+              ENOTCONN,
+              std::system_category(),
+              "EOF encountered during handshake"));
+            return make_exception_future(_error);
+        } else if (connected()) {
+            return make_ready_future<>();
         }
-        default:
-            _error = std::make_exception_ptr(ossl_error("Unhandled error code observed"));
-            return make_exception_future<>(_error);
+        try {
+            // Same function for clients or servers, however clients
+            // will be sending data first
+            auto n = SSL_do_handshake(_ssl.get());
+            if (n <= 0) {
+                auto ssl_error = SSL_get_error(_ssl.get(), n);
+                switch (ssl_error) {
+                case SSL_ERROR_NONE:
+                    // probably shouldn't have gotten here, but we're gtg
+                    break;
+                case SSL_ERROR_ZERO_RETURN:
+                    // peer has closed
+                    _eof = true;
+                    break;
+                case SSL_ERROR_WANT_WRITE:
+                case SSL_ERROR_WANT_READ: {
+                    // Always first check to see if there's any data to send.  Then wait
+                    // for data to be received.
+                    return maybe_perform_push_with_wait().then([this](bool) {
+                        return perform_pull().then(
+                          [this] { return do_handshake(); });
+                    });
+                    break;
+                }
```

Comment:
It feels strange to me that we can't handle these conditions independently.

Is it true that if we WANT_WRITE, a whole message is sent and the client will always (be expected to) respond?